x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/Evaluation.java

```diff
@@ -27,37 +34,67 @@
+
     /**
      * Returns the list of metrics to evaluate
      * @return list of metrics to evaluate
      */
     List<? extends EvaluationMetric> getMetrics();
 
+    default <T extends EvaluationMetric> List<T> initMetrics(@Nullable List<T> parsedMetrics, Supplier<List<T>> defaultMetricsSupplier) {
+        List<T> metrics = parsedMetrics == null ? defaultMetricsSupplier.get() : new ArrayList<>(parsedMetrics);
+        if (metrics.isEmpty()) {
+            throw ExceptionsHelper.badRequestException("[{}] must have one or more metrics", getName());
+        }
+        Collections.sort(metrics, Comparator.comparing(EvaluationMetric::getName));
+        return metrics;
+    }
+
     /**
      * Builds the search required to collect data to compute the evaluation result
      * @param userProvidedQueryBuilder User-provided query that must be respected when collecting data
      */
-    SearchSourceBuilder buildSearch(QueryBuilder userProvidedQueryBuilder);
-
-    /**
-     * Builds the search that verifies existence of required fields and applies user-provided query
-     * @param requiredFields fields that must exist
-     * @param userProvidedQueryBuilder user-provided query
-     */
-    default SearchSourceBuilder newSearchSourceBuilder(List<String> requiredFields, QueryBuilder userProvidedQueryBuilder) {
-        BoolQueryBuilder boolQuery = QueryBuilders.boolQuery();
-        for (String requiredField : requiredFields) {
-            boolQuery.filter(QueryBuilders.existsQuery(requiredField));
+    default SearchSourceBuilder buildSearch(QueryBuilder userProvidedQueryBuilder) {
+        ExceptionsHelper.requireNonNull(userProvidedQueryBuilder, "userProvidedQueryBuilder");
+        BoolQueryBuilder boolQuery =
+            QueryBuilders.boolQuery()
+                // Verify existence of required fields
+                .filter(QueryBuilders.existsQuery(getActualField()))
+                .filter(QueryBuilders.existsQuery(getPredictedField()))
+                // Apply user-provided query
+                .filter(userProvidedQueryBuilder);
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder().size(0).query(boolQuery);
+        for (EvaluationMetric metric : getMetrics()) {
+            // Fetch aggregations requested by individual metrics
+            List<AggregationBuilder> aggs = metric.aggs(getActualField(), getPredictedField());
+            aggs.forEach(searchSourceBuilder::aggregation);
         }
-        boolQuery.filter(userProvidedQueryBuilder);
-        return new SearchSourceBuilder().size(0).query(boolQuery);
+        return searchSourceBuilder;
     }
 
     /**
      * Processes {@link SearchResponse} from the search action
      * @param searchResponse response from the search action
      */
-    void process(SearchResponse searchResponse);
+    default void process(SearchResponse searchResponse) {
+        ExceptionsHelper.requireNonNull(searchResponse, "searchResponse");
```

Comment:
You can use `Objects.requireNonNull` here as this is the query is not provided directly from the user but from our code instead.

Reply:
Done.